File: algorithm/src/leetcode/RemoveNthNodeFromEndofList_19.java
package leetcode;

// #19 - Remove Nth Node From End of List
public class RemoveNthNodeFromEndofList_19 {
	static class ListNode {
		int val;
		ListNode next;
		ListNode() {}
		ListNode(int val) { this.val = val; }
		ListNode(int val, ListNode next) { this.val = val; this.next = next; }
	}

	public static void main(String[] args) {
		ListNode e = new ListNode(5);
		ListNode d = new ListNode(4, e);
		ListNode c = new ListNode(3, d);
		ListNode b = new ListNode(2, c);
		ListNode a = new ListNode(1, b);

		System.out.println("removeNthFromEnd(a, 2) = " + removeNthFromEnd(a, 1).val);
	}

	public static ListNode removeNthFromEnd(ListNode head, int n) {
		ListNode block = head;
		ListNode pointer = head;

		while(n > 0){
			block = block.next;
			if(block == null) {
				return head.next;
			}
			n--;
		}

		while(block.next != null){
			block = block.next;
			pointer = pointer.next;
		}

		// pointer.next 삭제
		pointer.next = pointer.next.next;

		return head;
	}
}

File: algorithm/src/leetcode/ruby/RemoveNthNodeFromEndOfList_19.rb
# 19 Remove Nth Node From End of List

class ListNode
  attr_accessor :val, :next
  def initialize(val = 0, _next = nil)
    @val = val
    @next = _next
  end
end

# @param {ListNode} head
# @param {Integer} n
# @return {ListNode}
def remove_nth_from_end(head, n)
  pointer = head
  block = head

  while n > 0
    block = block.next
    return head.next if block.nil?
    n -= 1
  end

  while !block.next.nil?
    block = block.next
    pointer = pointer.next
  end

  # pointer.next 삭제
  pointer.next = pointer.next.next

  head
end


h5 = ListNode.new(5)
h4 = ListNode.new(4, h5)
h3 = ListNode.new(3, h4)
h2 = ListNode.new(2, h3)
h1 = ListNode.new(1, h2)

n = 2
p remove_nth_from_end(h1, n)